### ISSUE TITLE
Server process progress over Socket.IO

### ISSUE THREAD

**Issue by ColdFire87** [miguelgrinberg/Flask-SocketIO#261]
Hi, I'm trying to display a progress bar in the front end for a process that happens on the server. I'm trying to send progress update messages over websockets but the problem that I am facing is that they all come together at the same time rather than as they are emitted.

I basically launch the process on the client side using:
`socket.emit('get:data', { input:'foo' });`
And then I listen for update messages:
`socket.on('update:progress', function (data){ console.log(data) });`
 
Then on the server:
`@socketio.on('get:data')
def get_data(input):
get_random()`

In the `get_random` function I generate some random data in a while loop and on each iteration I do `emit('update:progress', {'currentIteration': current_run})`

I would expect each `emit` call to fire the message over the socket immediately however it seems that the messages are buffered and the buffer is flushed when the `@socketio.on('get:data')` handler finishes.

I'm new to Python and Flask so I would appreciate if you could help me out with this issue.

Thanks in advance!

**Comment by miguelgrinberg** [comment]
Yes, this is a common mistake that is easy to address. You are probably using eventlet or gevent, correct?

What you need to keep in mind is that eventlet and gevent are coroutine frameworks. Any parallel processing only occurs if the threads release the CPU to allow the task scheduler to give CPU time to all concurrent threads. When you call `emit()`, there is a background worker thread from Flask-SocketIO that needs to do some work, and for that work to occur the long task needs to release the CPU.

A task switch can occur implicitly when a thread calls an I/O function. If your long function does not do I/O, then you have to explicitly release the CPU every once in a while. The way to explicitly release CPU in eventlet, for example, is to call `eventlet.sleep()`. This will let other waiting tasks run, and eventually control will return back to your function. This is something that you should do often, for example once in each loop iteration, if your function does some work in a loop. If you don't release the CPU, all the `emit()` calls that you make will be queued and occur once the CPU is release, which based on your description happens when your long task ends.

Hope this makes sense. Let me know if anything here isn't clear.

**Comment by ColdFire87** [comment]
Wow! Thanks Miguel! You are a life saver! It worked :D
Thanks for writing this module!

**Comment by jddeltoro** [comment]
I have the same problem but just using node with socket.io, which will be the equivalent of eventlet.sleep()?

**Comment by miguelgrinberg** [comment]
@jddeltoro I don't think there is an equivalent to `sleep()` in JavaScript. In the browser, you can probably get something similar implemented using `setTimeout()`. If you are using Node on the server, maybe `process.nextTick()`.

**Comment by jddeltoro** [comment]
@miguelgrinberg thanks for quick response. I will briefly explain the problem that I can not solve.
this is my super simplified code.
```
//Server side function
functionThatShouldEmitInTime(data){
	readFileWithCallback(data, function(dataUpdated){
		readFileWithCallback(dataUpdated, function(dataEvenMoreUpdated){
			var w = dataEvenMoreUpdated.w;
			for (var c in w) {
				for (var e in c){
					var to_be_filled;
					//Do stuff
					for (var a in e) {
						//do stuff with to_be_filled
						for(var pr in a){
							//do stuff with to_be_filled
							for(var pe i pr){
								//do stuff with to_be_filled
							}
						}
					}
					writeFileAsync(to_be_filled);
					console.log(to_be_filled.name + ' is being written');
					socket.emit('file:processed', 'type_of_file');
				}
			}
		});
	});
};
```
This works fine, only that the emits, are sent when all loops end. Therefore the progress in the client is displayed instantly and at the end of the operation. Not gradually as it goes happening.
I used the variant that you indicate to me as I show below, but the result is the same.

```
//Server side function
functionThatShouldEmitInTime(data){
	readFileWithCallback(data, function(dataUpdated){
		readFileWithCallback(dataUpdated, function(dataEvenMoreUpdated){
			var c = dataEvenMoreUpdated.c;
			for (var c in w) {
				for (var e in c){
					var to_be_filled;
					//Do stuff
					for (var a in e) {
						//do stuff with to_be_filled
						for(var pr in a){
							//do stuff with to_be_filled
							for(var pe i pr){
								//do stuff with to_be_filled
							}
						}
					}
					writeFileAsync(to_be_filled);
					console.log(to_be_filled.name + ' is being written');
					process.nextTick(force_emit);
				}
			}
		});
	});
};

function force_emit() {
    socket.emit('file:processed', 'type_of_file');
}
```

I also read and tried all the information in this StackOverflow posts but without success:
[https://stackoverflow.com/questions/29852972/sending-multiple-emits-inside-a-loop](url)
[https://stackoverflow.com/questions/25309886/nodejs-socket-io-emit-inside-function-loop](url)
[https://stackoverflow.com/questions/43086991/socket-io-emit-in-loop-until-client-responds](url)
[https://stackoverflow.com/questions/30898258/socket-io-inside-of-loop-doesnt-work](url)

Some ideas that could help me. Thanks in advanced.
Jorge

**Comment by miguelgrinberg** [comment]
@jddeltoro Not sure how to help. Async servers do not do well when you have a lot of CPU intensive work. Offloading the CPU work to subprocess might help, you really need to make sure your loops do not block for too long.

**Comment by jddeltoro** [comment]
@miguelgrinberg Anyway, thank you for your time. I'll tell you here if I can solve it.

**Comment by inkeytech** [comment]
Hello jddeltoro,

Did you ever have any luck solving this?

I'm experiencing the same issue.

I thought using an asynchronous each loop would help, but it didn't.

Regards,

Ian Watson

**Comment by jddeltoro** [comment]
Hi @inkeytech 

Yes i did.

Basically I used  async library. I leave a socket listening on the client "filesProcesed", each time a result is obtained from the soap service, the result is written on the disk, only then the emit GENERAL_SOCKET.emit('filesProcesed', 'service_response'); is executed and then callback() function allows to iterate to the next services request and repeat the process.

Hope this can help you.

`async.eachOfSeries(objCollection, function(item, key, callback) {
                client.ProcessRequest(item.args, function(err, result) {
                    var timeOfRequest = new Date().toLocaleString();
                    console.log(key + " -Service Responses Requested for: " + item.filename + " at: " + timeOfRequest);
                    var filename = item.rutCompany + "_" + item.rutEmployee;
                    file_procesor.write_permisions_xml_file(MANAGER.MODULES.MODULE_PERMISIONS_MIGRATIONS.RESPONSE_PERMISIONS_FILE_DIRECTORY, filename, result);
                    GENERAL_SOCKET.emit('filesProcesed', 'service_response');
                    GENERAL_SOCKET.emit('console_log', key + " -Service Responses Requested for: " + item.filename + " at: " + timeOfRequest);
                    callback();
                });
            });`

**Comment by inkeytech** [comment]
Hello @jddeltoro ,

Thank you very much for the quick and detailed response.

I'll be trying this now, and will post my results.

**Comment by inkeytech** [comment]
@jddeltoro,

Your suggestion worked perfectly.

My code was almost an exact replica of yours, except that I was using **async.each** AND NOT **async.eachSeries**, or **async.eachOfSeries**.

That is what makes all the difference.

Initially, using **async.eachSeries** resulted in a '_Maximum call stack size exceeded_' error after about 3000 iterations.

Using **async.setImmediate**, the suggested solution, solved the issue.

Thanks again,

Ian

**Comment by jddeltoro** [comment]
@inkeytech, glad of help you.

I process a large amount of data, and it happens that when these exceed 3000, the code stops, but I can not catch the exception, it is possible that I am dealing with this same error.

I solved it by executing batches of less than 3000 elements, but it would be very useful to execute the process completely
Could you tell me a little more in detail how you solved it?

Thank you.
Jorge

**Comment by inkeytech** [comment]
Hello @jddeltoro ,

I suspect you are experiencing exactly the same issue I had.

I found the answer somewhere on this site: http://caolan.github.io/async/.

When I looked for the exact snippet to reference in my response to you, I couldn't find it.

I'm only starting to learn the node.js event loop, so I cannot give you any technical reasons, just some anecdotal stuff.

If you use **each**, instead of **eachSeries** (**eachOf** instead of **eachOfSeries**) the operation runs a lot faster and will not break at 3000.

Is there a specific reason you use **eachOfSeries**? If there is no valid reason you could possibly use **eachOf**, but then your websocket calls will probably back up.

The ONLY reason I NOW use **eachSeries** is to allow my socket.io emits to work without backing up.

This is only to provide a good UX, and I might still scrap it.

TL;DR:

Here is code that hopefully shows what I'm doing:

If the code inside **async.eachSeries** is sync:

```
async.eachSeries(arNewContacts, function(contact, next){
	// do stuff with contact
	contact.calcValue = contact.firstName + contact.surname;
	globalArray.push(contact);
	
	async.setImmediate(function(){
		next();
	});
})

```

If the code inside **async.eachSeries** is async:

```
async.eachSeries(arNewContacts, function(contact, next){
	// do stuff with contact
	contact.calcValue = contact.firstName + contact.surname;
	fs.writeFile('helloworld.txt', contact, function(err){
		async.setImmediate(function(){
			next();
		});
	});
})

```
I recall reading that you need not use async.setImmediate to call your callback every single time, but only once every 1000 operations or so.

eg:

if (counter > 1000){
                async.setImmediate(function(){
			next();
		});
} else {
              next();
}

I hope this helps.

Ian Watson

**Comment by jddeltoro** [comment]
@inkeytech Hey men, thank you for the tip, i had to stop this project some time ago, and just now continue where i left, I followed your recomendation and everything worked smoothly.
